### PR TITLE
Added infrastructure for varnish purging and a test

### PR DESCRIPTION
**Pull request by Pchelolo**
Moving Varnish purging to change-propagation.

cc @wikimedia/services

### CODE REVIEW

**Comment by gwicke** [comment]
:+1: on the code change from me.

How (in which order) should we deploy this?

**Comment by Pchelolo** [comment]
@gwicke I think that these PRs that add use-cases for change-prop should go one-by-one.

**Comment by gwicke** [comment]
@Pchelolo, another concern for migrating purging is executing those purges twice. Can we target only a subset of purges at first?

**Comment by Pchelolo** [comment]
@gwicke For that we will add a more targeted match in puppet config, but here we need the whole thing right away

**Comment by d00rman** [comment]
Aren't purges supposed to be handled by a separate service (as we discussed with the Performance team) that listens to the `resource_changed` topic?

**Comment by gwicke** [comment]
@d00rman, my understanding is that purge handling in the change propagation service would be a temporary solution, until the separate purge service arrives & takes over. I would not be surprised if that will take some time, and will require some more thinking around XKey & timing of purges.

**Comment by d00rman** [comment]
This doesn't look right to me. The service purges RESTBase content as soon as MW emits a purge event? There is a strong possibility there will be a window of time where the content has been purged, but not yet re-rendered. The correct way would be for it to match `api/rest_v1` routes in the `resource_change` (which are anyway emitted by RESTBase) and then purge that content.

**Comment by Pchelolo** [comment]
@d00rman but that's exactly what's happening (https://github.com/wikimedia/change-propagation/pull/16/files#diff-541b6e195e9da580fc0bd42761ee761fR85)

File: lib/rule_executor.js [hunk 1]
```diff
@@ -199,6 +199,7 @@ class RuleExecutor {
                         return this._exec(message)
                         .catch((e) => {
                             if (this.rule.shouldRetry(e)) {
+                                this.log('warn/retry', e);
```
Comment (d00rman) [inline lib/rule_executor.js]: Let's not do that. When part of the infra is down or a similar issue occurs, we'll have a gazillion log entries.

File: config.example.wikimedia.yaml [hunk 2]
```diff
@@ -68,11 +74,25 @@ services:
...
                         # Don't encode title since it should be already encoded
                         uri: '{{match.meta.uri[1]}}://{{message.meta.domain}}/api/rest_v1/page/definition/{{match.meta.uri[2]}}'
                         headers:
                           cache-control: no-cache
+
+                    purge_varnish:
+                      topic: resource_change
+                      match:
+                        meta:
+                          uri: '/^https?:\/\/[a-zA-Z0-9:\.]+\/api\/rest_v1\/(.+)$/'
```
Comment (d00rman) [inline config.example.wikimedia.yaml]: Domain regexes are really tricky. Let's use `[^\/]+` instead

**Comment by Pchelolo** [comment]
@d00rman I've addressed your comments

**Pull request merged**